internal/digraph/scheduler/scheduler.go

```diff
@@ -587,3 +587,60 @@ func (sc *Scheduler) isSucceed(g *ExecutionGraph) bool {
 func (sc *Scheduler) isTimeout(startedAt time.Time) bool {
 	return sc.timeout > 0 && time.Since(startedAt) > sc.timeout
 }
+
+// handleNodeRetry handles the retry logic for a node based on exit codes and retry policy
+func (sc *Scheduler) handleNodeRetry(ctx context.Context, node *Node, execErr error) (shouldRetry bool) {
+	var exitCode int
+	var exitCodeFound bool
+
+	// Try to extract exit code from different error types
+	if exitErr, ok := execErr.(*exec.ExitError); ok {
+		exitCode = exitErr.ExitCode()
+		exitCodeFound = true
+		logger.Debug(ctx, "Found ExitError", "error", execErr, "exitCode", exitCode)
+	} else {
+		// Try to parse exit code from error string
+		errStr := execErr.Error()
+		if strings.Contains(errStr, "exit status") {
+			// Parse "exit status N" format
+			parts := strings.Split(errStr, " ")
+			if len(parts) > 2 {
+				if code, err := strconv.Atoi(parts[2]); err == nil {
+					exitCode = code
+					exitCodeFound = true
+					logger.Debug(ctx, "Parsed exit code from error string", "error", errStr, "exitCode", exitCode)
+				}
+			}
+		} else if strings.Contains(errStr, "signal:") {
+			// Handle signal termination
+			exitCode = -1
+			exitCodeFound = true
+			logger.Debug(ctx, "Process terminated by signal", "error", errStr)
+		}
+	}
+
+	if !exitCodeFound {
+		logger.Debug(ctx, "Could not determine exit code", "error", execErr, "errorType", fmt.Sprintf("%T", execErr))
+		// Default to exit code 1 if we can't determine the actual code
+		exitCode = 1
+	}
+
+	shouldRetry = node.retryPolicy.ShouldRetry(exitCode)
+	logger.Debug(ctx, "Checking retry policy", "exitCode", exitCode, "allowedCodes", node.retryPolicy.ExitCodes, "shouldRetry", shouldRetry)
+
+	if !shouldRetry {
+		// finish the node with error
+		node.data.SetStatus(NodeStatusError)
+		node.data.MarkError(execErr)
+		sc.setLastError(execErr)
+		return false
+	}
```

Comment:
Would it be possible to reuse the same logic in the switch statement below? In some cases, a step should be marked as successful even if it fails.

https://github.com/dagu-org/dagu/blob/2ab2fab4636f69f0dc1e7c0d913673f1f143de0f/internal/digraph/scheduler/scheduler.go#L203-L214

Perhaps we could check the return value from the `handleNodeRetry` fucntion and use fallthrough when the return value is false. What do you think?